src/DoctrineMessageRepository/DoctrineMessageRepository.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+namespace EventSauce\MessageRepository\DoctrineMessageRepository;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use EventSauce\EventSourcing\AggregateRootId;
+use EventSauce\EventSourcing\Header;
+use EventSauce\EventSourcing\Message;
+use EventSauce\EventSourcing\MessageRepository;
+use EventSauce\EventSourcing\Serialization\MessageSerializer;
+use EventSauce\EventSourcing\UnableToPersistMessages;
+use EventSauce\EventSourcing\UnableToRetrieveMessages;
+use Generator;
+use Ramsey\Uuid\Uuid;
+use Throwable;
+
+use function count;
+use function implode;
+use function json_decode;
+use function json_encode;
+use function sprintf;
+
+class DoctrineMessageRepository implements MessageRepository
+{
+    private int $jsonEncodeOptions = 0;
+
+    public function __construct(
+        private Connection $connection,
+        private MessageSerializer $serializer,
+        private TableSchema $tableSchema,
+        private UuidSerializer $uuidSerializer,
+    ) {}
+
+    public function persist(Message ...$messages): void
+    {
+        if (count($messages) === 0) {
+            return;
+        }
+
+        try {
+            $this->connection->transactional(fn () => $this->persistMessages($messages));
```

Comment:
The previous implementation use string interpolation to generate a single `INSERT` query with many values. While that implementation may have been (slightly) more performant, I feel that running multiple `INSERT` statements in a transaction is just as safe and easier to understand.

Reply:
Was there a reason why you rewrote this with the query builder? I prefer to have one statement. At work we use these implementations as a big highway for events, the latency for the DB is multiplied by the number of events, which makes it unpredictable and thus harder to account for capacity planning.

Reply:
As my comment above says, I felt that a single transaction with multiple `INSERT` statements would be nearly as performant and **much** easier to understand from a code standpoint. If you feel that a single query represents a significant enough performance benefit, I am happy to return the previous implementation.

Reply:
I'd prefer that. The O(n) latency is not great for apps at the scale where Mollie operates.

Reply:
Do you feel that it is important to use `:named` parameters for the query, as opposed to simple `?` placeholders? The former results in quite a few string concatenation gymnastics, which was part of the reason I switched to using `Connection::insert`.

Reply:
I do not have a particularly strong opinion about that. There might be a param binding type difference tho, right?

Reply:
I implemented named parameters but I don't think there is a strong reason to do it, and it definitely makes generating the query more complicated. I think I'll make a separate commit that uses un-named parameters to see if it cleans things up.